File: scripts/spike-pos-caps-repair-roi.mjs
#!/usr/bin/env node
/**
 * Spike ROI — bulk G2 (3 passes total, not per-file).
 *   node scripts/spike-pos-caps-repair-roi.mjs
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT } from './lib/loadEnv.mjs';
import { collectStringsFromBatch, runPosCapsBulk } from './lib/germanCapsGate.mjs';
import { applyGermanCapsNormalize } from './lib/germanCapsNormalize.mjs';

const HOLDOUT_DIR = path.join(ROOT, 'batches/ready/lesen');
const DRYRUN_PATH = path.join(ROOT, 'batches/ready/PHASE1-G2-DRYRUN.json');
const OUT_JSON = path.join(ROOT, 'batches/ready/gate-logs/spike-pos-caps-repair-roi.json');
const OUT_MD = path.join(ROOT, 'batches/ready/gate-logs/SPIKE-POS-CAPS-REPAIR-ROI.md');

const PURE_REPAIR_REASONS = new Set([
  'adj_before_noun', 'quantifier_capitalized', 'adj_after_prep', 'adv_capitalized',
  'adv_after_pronoun', 'zu_adv_capitalized', 'lexicon_nn', 'modal_noun_object',
  'lexicon_after_adj', 'lexicon_override_tag', 'double_pass_after_prep', 'adv_before_verb',
]);

const RISKY_REPAIR_REASONS = new Set([
  'verb_census_no_finite', 'prose_strict_homograph', 'modal_final_infinitive',
]);

function listHoldoutFiles() {
  return fs.readdirSync(HOLDOUT_DIR)
    .filter((f) => f.endsWith('.json') && /lesen-t\d/i.test(f))
    .sort();
}

function loadAllItems(files) {
  const items = [];
  const batches = new Map();
  for (const file of files) {
    const batch = JSON.parse(fs.readFileSync(path.join(HOLDOUT_DIR, file), 'utf8'));
    batches.set(file, batch);
    for (const { field, text } of collectStringsFromBatch(batch)) {
      items.push({ id: `${file}::${field}::${items.length}`, file, field, text });
    }
  }
  return { items, batches };
}

function runG2(items) {
  const bulk = runPosCapsBulk(items, { timeoutMs: 600_000, chunkSize: 200 });
  if (bulk.skipped) throw new Error(bulk.warning);
  const idToMeta = new Map(items.map((it) => [it.id, it]));
  return (bulk.findings || []).map((f) => ({
    ...f,
    file: idToMeta.get(String(f.id))?.file || f.file,
    field: f.field || idToMeta.get(String(f.id))?.field,
  }));
}

function escapeRe(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function applyMechanicalFix(text, finding) {
  const { word, type, context } = finding;
  const target = type === 'wrong_capitalized'
    ? word[0].toLowerCase() + word.slice(1)
    : word[0].toUpperCase() + word.slice(1);
  if (word === target) return text;
  const ctx = String(context || '');
  const idx = text.indexOf(ctx);
  if (idx >= 0 && ctx.includes(word)) {
    const wIdx = ctx.indexOf(word);
    const abs = idx + wIdx;
    return text.slice(0, abs) + target + text.slice(abs + word.length);
  }
  const re = new RegExp(`(?<![A-Za-zÄÖÜäöüß-])${escapeRe(word)}(?![A-Za-zÄÖÜäöüß-])`);
  const m = re.exec(text);
  if (!m) return text;
  return text.slice(0, m.index) + target + text.slice(m.index + word.length);
}

function setFieldText(batch, field, newText) {
  const b = structuredClone(batch);
  const pm = field.match(/^passages\[(\d+)\]\.(.+)$/);
  if (pm) {
    const p = b.passages[Number(pm[1])];
    const rest = pm[2];
    if (rest === 'text') p.text = newText;
    else if (rest === 'title') p.title = newText;
    else if (rest === 'transcript') p.transcript = newText;
    else if (rest.startsWith('ads[')) p.ads[Number(rest.match(/ads\[(\d+)\]/)?.[1])] = newText;
    else if (rest.startsWith('audio[')) p.audio[Number(rest.match(/audio\[(\d+)\]/)?.[1])].text = newText;
    return b;
  }
  const qm = field.match(/^questions\[(\d+)\]\.(.+)$/);
  if (qm) {
    const q = b.questions[Number(qm[1])];
    const rest = qm[2];
    if (rest === 'question') q.question = newText;
    else if (rest === 'signText') q.signText = newText;
    else if (rest === 'explanation') q.explanation = newText;
    else if (rest === 'statement') q.statement = newText;
    else if (rest.startsWith('options[')) {
      const oi = Number(rest.match(/options\[(\d+)\]/)?.[1]);
      if (typeof q.options[oi] === 'string') q.options[oi] = newText;
      else if (q.options[oi]) q.options[oi].text = newText;
    } else if (rest.startsWith('matchLabels[')) {
      q.matchLabels[Number(rest.match(/matchLabels\[(\d+)\]/)?.[1])] = newText;
    }
  }
  return b;
}

function applyMechanicalToBatches(batches, findings, items) {
  const textById = new Map(items.map((it) => [it.id, it.text]));
  const findingsById = new Map();
  for (const f of findings) {
    const id = f.id || `${f.file}::${f.field}`;
    if (!findingsById.has(id)) findingsById.set(id, []);
    findingsById.get(id).push(f);
  }
  const out = new Map(batches);
  for (const [id, flist] of findingsById) {
    let text = textById.get(String(id));
    if (text == null) continue;
    for (const f of flist) text = applyMechanicalFix(text, f);
    const meta = items.find((it) => it.id === String(id));
    if (!meta) continue;
    out.set(meta.file, setFieldText(out.get(meta.file), meta.field, text));
  }
  return out;
}

function classify(f) {
  if (PURE_REPAIR_REASONS.has(f.reason)) return 'repair-by-type';
  if (RISKY_REPAIR_REASONS.has(f.reason)) return 'repair-by-type-risky';
  return 'needs-review';
}

function countBy(arr, fn) {
  const m = {};
  for (const x of arr) m[fn(x)] = (m[fn(x)] || 0) + 1;
  return m;
}

function pct(n, d) {
  return d ? Math.round((n / d) * 1000) / 10 : 0;
}

function diffFindings(before, after) {
  const key = (f) => `${f.file}::${f.field}::${f.word}::${f.reason}`;
  const afterSet = new Set(after.map(key));
  const beforeSet = new Set(before.map(key));
  return {
    resolved: before.filter((f) => !afterSet.has(key(f))),
    added: after.filter((f) => !beforeSet.has(key(f))),
  };
}

async function main() {
  const dryrun = JSON.parse(fs.readFileSync(DRYRUN_PATH, 'utf8'));
  const files = listHoldoutFiles();
  const { items, batches } = loadAllItems(files);

  console.log(`Bulk G2 pass 1/3 (before) — ${items.length} fields…`);
  const before = runG2(items);
  console.log(`Findings before: ${before.length} (dryrun: ${dryrun.summary.beforeFindings})`);

  const byClass = { 'repair-by-type': [], 'repair-by-type-risky': [], 'needs-review': [] };
  for (const f of before) byClass[classify(f)].push(f);

  console.log('Applying mechanical repair + bulk G2 pass 2/3…');
  const mechBatches = applyMechanicalToBatches(batches, before, items);
  const mechItems = [];
  for (const file of files) {
    for (const { field, text } of collectStringsFromBatch(mechBatches.get(file))) {
      mechItems.push({ id: `${file}::${field}::${mechItems.length}`, file, field, text });
    }
  }
  const afterMech = runG2(mechItems);
  const mechDiff = diffFindings(before, afterMech);
  console.log(`After mechanical: ${afterMech.length} (resolved ${mechDiff.resolved.length}, added ${mechDiff.added.length})`);

  console.log('Lists normalize + bulk G2 pass 3/3…');
  const listItems = [];
  for (const file of files) {
    const { batch } = applyGermanCapsNormalize(batches.get(file), { decapOnly: false });
    for (const { field, text } of collectStringsFromBatch(batch)) {
      listItems.push({ id: `${file}::${field}::${listItems.length}`, file, field, text });
    }
  }
  const afterLists = runG2(listItems);
  const listsDiff = diffFindings(before, afterLists);
  console.log(`After lists: ${afterLists.length} (dryrun: ${dryrun.summary.afterFindings})`);

  const report = {
    generatedAt: new Date().toISOString(),
    holdoutFiles: files.length,
    fields: items.length,
    dryrunRef: { before: dryrun.summary.beforeFindings, after: dryrun.summary.afterFindings },
    measured: {
      beforeFindings: before.length,
      afterMechanical: afterMech.length,
      afterListsNormalize: afterLists.length,
    },
    classification: {
      repairByTypePure: byClass['repair-by-type'].length,
      repairByTypeRisky: byClass['repair-by-type-risky'].length,
      needsReview: byClass['needs-review'].length,
      purePct: pct(byClass['repair-by-type'].length, before.length),
      riskyPct: pct(byClass['repair-by-type-risky'].length, before.length),
      byReasonBefore: countBy(before, (f) => f.reason),
      pureByReason: countBy(byClass['repair-by-type'], (f) => f.reason),
      riskyByReason: countBy(byClass['repair-by-type-risky'], (f) => f.reason),
    },
    mechanicalSimulation: {
      resolvedCount: mechDiff.resolved.length,
      remainingCount: afterMech.length,
      addedCount: mechDiff.added.length,
      resolvedPct: pct(mechDiff.resolved.length, before.length),
      addedFindings: mechDiff.added.slice(0, 20),
      unresolvedSample: afterMech.slice(0, 25),
    },
    listsComparison: {
      resolvedCount: listsDiff.resolved.length,
      remainingCount: afterLists.length,
      resolvedPct: pct(listsDiff.resolved.length, before.length),
      deltaMechVsLists: afterMech.length - afterLists.length,
    },
  };

  fs.mkdirSync(path.dirname(OUT_JSON), { recursive: true });
  fs.writeFileSync(OUT_JSON, `${JSON.stringify(report, null, 2)}\n`);
  fs.writeFileSync(OUT_MD, renderMd(report));
  console.log(`Escrito: ${OUT_MD}`);
}

function renderMd(r) {
  const purePlusRisky = r.classification.repairByTypePure + r.classification.repairByTypeRisky;
  return [
    '# Spike ROI — pos-caps repair mecánico',
    '',
    `**Fecha:** ${r.generatedAt}`,
    `**Corpus:** ${r.holdoutFiles} archivos, ${r.fields} campos`,
    '',
    '## 1. Clasificación on-paper',
    '',
    '| Clase | N | % |',
    '|---|---:|---:|',
    `| repair-by-type (puro) | ${r.classification.repairByTypePure} | ${r.classification.purePct}% |`,
    `| repair-by-type-risky | ${r.classification.repairByTypeRisky} | ${r.classification.riskyPct}% |`,
    `| **Total candidatos mecánicos** | **${purePlusRisky}** | **${pct(purePlusRisky, r.measured.beforeFindings)}%** |`,
    '',
    '## 2. Simulación',
    '',
    '| Estrategia | Findings tras fix | Resueltos | addedFindings |',
    '|---|---:|---:|---:|',
    `| Baseline (sin fix) | ${r.measured.beforeFindings} | 0 | 0 |`,
    `| **Repair mecánico naive** | ${r.measured.afterMechanical} | ${r.mechanicalSimulation.resolvedCount} (${r.mechanicalSimulation.resolvedPct}%) | ${r.mechanicalSimulation.addedCount} |`,
    `| Lists normalize (v3.2) | ${r.measured.afterListsNormalize} | ${r.listsComparison.resolvedCount} (${r.listsComparison.resolvedPct}%) | — |`,
    '',
    `**Delta mecánico vs lists:** ${r.listsComparison.deltaMechVsLists} findings restantes (positivo = mecánico peor).`,
  ].join('\n');
}

main().catch((e) => { console.error(e); process.exit(1); });

File: scripts/spike-pos-caps-repair-roi.py
#!/usr/bin/env python3
"""Spike ROI: mechanical repair simulation on holdout (single spaCy load)."""
from __future__ import annotations

import importlib.util
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HOLDOUT = ROOT / "batches/ready/lesen"
DRYRUN = ROOT / "batches/ready/PHASE1-G2-DRYRUN.json"
OUT_JSON = ROOT / "batches/ready/gate-logs/spike-pos-caps-repair-roi.json"
OUT_MD = ROOT / "batches/ready/gate-logs/SPIKE-POS-CAPS-REPAIR-ROI.md"

PURE = {
    "adj_before_noun", "quantifier_capitalized", "adj_after_prep", "adv_capitalized",
    "adv_after_pronoun", "zu_adv_capitalized", "lexicon_nn", "modal_noun_object",
    "lexicon_after_adj", "lexicon_override_tag", "double_pass_after_prep", "adv_before_verb",
}
RISKY = {"verb_census_no_finite", "prose_strict_homograph", "modal_final_infinitive"}


def load_pc():
    spec = importlib.util.spec_from_file_location("pc", ROOT / "scripts/pos-caps-check.py")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    mod.load_nlp()
    mod.load_lexicon()
    return mod


def collect_strings(batch: dict) -> list[tuple[str, str]]:
    out = []
    for pi, p in enumerate(batch.get("passages") or []):
        if isinstance(p.get("text"), str):
            out.append((f"passages[{pi}].text", p["text"]))
        if isinstance(p.get("title"), str):
            out.append((f"passages[{pi}].title", p["title"]))
        for ai, ad in enumerate(p.get("ads") or []):
            if isinstance(ad, str):
                out.append((f"passages[{pi}].ads[{ai}]", ad))
    for qi, q in enumerate(batch.get("questions") or []):
        for key in ("question", "signText", "explanation", "statement"):
            if isinstance(q.get(key), str):
                out.append((f"questions[{qi}].{key}", q[key]))
        for oi, opt in enumerate(q.get("options") or []):
            t = opt if isinstance(opt, str) else opt.get("text")
            if isinstance(t, str):
                out.append((f"questions[{qi}].options[{oi}]", t))
    return out


def analyze_fields(pc, fields: list[tuple[str, str]], file: str, regime: str = "PROSE"):
    findings = []
    for ti, (field, text) in enumerate(fields):
        fs, _ = pc.analyze_text(text, ti, field, regime)
        for f in fs:
            f = dict(f)
            f["file"] = file
            f["field"] = field
            findings.append(f)
    return findings


def mechanical_fix(text: str, finding: dict) -> str:
    word = finding["word"]
    ftype = finding["type"]
    target = word[0].lower() + word[1:] if ftype == "wrong_capitalized" else word[0].upper() + word[1:]
    if word == target:
        return text
    ctx = finding.get("context") or ""
    idx = text.find(ctx)
    if idx >= 0 and word in ctx:
        abs_i = idx + ctx.index(word)
        return text[:abs_i] + target + text[abs_i + len(word):]
    m = re.search(r"(?<![A-Za-zÄÖÜäöüß-])" + re.escape(word) + r"(?![A-Za-zÄÖÜäöüß-])", text)
    if not m:
        return text
    return text[: m.start()] + target + text[m.end() :]


def apply_mech_to_fields(fields: list[tuple[str, str]], findings: list[dict]) -> list[tuple[str, str]]:
    by_field: dict[str, list[dict]] = {}
    for f in findings:
        by_field.setdefault(f["field"], []).append(f)
    out = []
    for field, text in fields:
        t = text
        for f in by_field.get(field, []):
            t = mechanical_fix(t, f)
        out.append((field, t))
    return out


def fkey(f: dict) -> str:
    return f"{f['file']}::{f['field']}::{f['word']}::{f['reason']}"


def classify(reason: str) -> str:
    if reason in PURE:
        return "repair-by-type"
    if reason in RISKY:
        return "repair-by-type-risky"
    return "needs-review"


def main():
    pc = load_pc()
    dryrun = json.loads(DRYRUN.read_text(encoding="utf-8"))
    files = sorted(p.name for p in HOLDOUT.glob("lesen-t*.json"))

    before: list[dict] = []
    file_fields: dict[str, list[tuple[str, str]]] = {}
    for fn in files:
        batch = json.loads((HOLDOUT / fn).read_text(encoding="utf-8"))
        fields = collect_strings(batch)
        file_fields[fn] = fields
        before.extend(analyze_fields(pc, fields, fn))

    print(f"Before: {len(before)} (ref {dryrun['summary']['beforeFindings']})", file=sys.stderr)

    by_class = {"repair-by-type": 0, "repair-by-type-risky": 0, "needs-review": 0}
    for f in before:
        by_class[classify(f["reason"])] += 1

    after_mech: list[dict] = []
    for fn in files:
        fields = file_fields[fn]
        bf = [x for x in before if x["file"] == fn]
        mfields = apply_mech_to_fields(fields, bf)
        after_mech.extend(analyze_fields(pc, mfields, fn))

    before_set = {fkey(f) for f in before}
    after_mech_set = {fkey(f) for f in after_mech}
    resolved = [f for f in before if fkey(f) not in after_mech_set]
    added = [f for f in after_mech if fkey(f) not in before_set]

    print(f"After mech: {len(after_mech)} resolved={len(resolved)} added={len(added)}", file=sys.stderr)

    # Lists normalize via node — use dryrun ref for after lists
    after_lists_ref = dryrun["summary"]["afterFindings"]
    lists_resolved = len(before) - after_lists_ref

    report = {
        "generatedAt": __import__("datetime").datetime.utcnow().isoformat() + "Z",
        "holdoutFiles": len(files),
        "dryrunRef": dryrun["summary"],
        "classification": {
            **by_class,
            "purePct": round(1000 * by_class["repair-by-type"] / len(before)) / 10,
            "riskyPct": round(1000 * by_class["repair-by-type-risky"] / len(before)) / 10,
            "mechanicalCandidatePct": round(1000 * (by_class["repair-by-type"] + by_class["repair-by-type-risky"]) / len(before)) / 10,
            "byReasonBefore": dryrun["summary"]["beforeByReason"],
        },
        "mechanicalSimulation": {
            "before": len(before),
            "after": len(after_mech),
            "resolved": len(resolved),
            "resolvedPct": round(1000 * len(resolved) / len(before)) / 10,
            "added": len(added),
            "addedSample": added[:15],
            "unresolvedSample": after_mech[:20],
        },
        "listsNormalize": {
            "afterRef": after_lists_ref,
            "resolvedRef": lists_resolved,
            "resolvedPctRef": round(1000 * lists_resolved / len(before)) / 10,
            "deltaMechVsLists": len(after_mech) - after_lists_ref,
        },
    }

    OUT_JSON.parent.mkdir(parents=True, exist_ok=True)
    OUT_JSON.write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")

    md = f"""# Spike ROI — pos-caps repair mecánico

**Fecha:** {report['generatedAt']}
**Corpus:** {len(files)} archivos holdout
**Método:** simulación Python (misma lógica G2, repair mecánico token-a-token)

## 1. Clasificación on-paper

| Clase | N | % |
|---|---:|---:|
| repair-by-type (puro) | {by_class['repair-by-type']} | {report['classification']['purePct']}% |
| repair-by-type-risky | {by_class['repair-by-type-risky']} | {report['classification']['riskyPct']}% |
| **Total candidatos mecánicos** | **{by_class['repair-by-type'] + by_class['repair-by-type-risky']}** | **{report['classification']['mechanicalCandidatePct']}%** |

## 2. Simulación

| Estrategia | Tras fix | Resueltos | addedFindings |
|---|---:|---:|---:|
| Baseline | {len(before)} | 0 | 0 |
| Repair mecánico naive | {len(after_mech)} | {len(resolved)} ({report['mechanicalSimulation']['resolvedPct']}%) | {len(added)} |
| Lists normalize (dryrun ref) | {after_lists_ref} | {lists_resolved} ({report['listsNormalize']['resolvedPctRef']}%) | 3 net new in dryrun |

**Delta mecánico vs lists:** {report['listsNormalize']['deltaMechVsLists']} findings restantes.
"""
    OUT_MD.write_text(md, encoding="utf-8")
    print(f"Wrote {OUT_MD}", file=sys.stderr)
    print(json.dumps(report["mechanicalSimulation"], indent=2))
    print(json.dumps(report["classification"], indent=2))


if __name__ == "__main__":
    main()
